Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -358,91 +377,129 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
         }
 
         settings.on_message_complete = { parser in
-            let handler = evacuateHTTPDecoder(parser)
-
+            let ctx = evacuateChannelHandlerContext(parser)
+            let handler = ctx.handler as! AnyHTTPDecoder
             handler.state.complete(state: handler.state.dataAwaitingState)
             handler.state.dataAwaitingState = .messageBegin
 
             let trailers = handler.state.currentHeaders.isEmpty ? nil : HTTPHeaders(buffer: handler.state.cumulationBuffer!, headers: handler.state.currentHeaders)
             handler.state.currentHeaders.removeAll(keepingCapacity: true)
             switch handler {
             case let handler as HTTPRequestDecoder:
-                handler.pendingInOut.append(handler.wrapInboundOut(HTTPServerRequestPart.end(trailers)))
+                ctx.fireChannelRead(handler.wrapInboundOut(HTTPServerRequestPart.end(trailers)))
             case let handler as HTTPResponseDecoder:
-                handler.pendingInOut.append(handler.wrapInboundOut(HTTPClientResponsePart.end(trailers)))
+                ctx.fireChannelRead(handler.wrapInboundOut(HTTPClientResponsePart.end(trailers)))
             default:
                 fatalError("the impossible happened: handler neither a HTTPRequestDecoder nor a HTTPResponseDecoder which should be impossible")
             }
             return 0
         }
     }
 
-    public func decoderRemoved(ctx: ChannelHandlerContext) {
-        // Remove the stored reference to ChannelHandlerContext
-        parser.data = UnsafeMutableRawPointer(bitPattern: 0x0000deadbeef0000)
-
-        // Set the callbacks to nil as we dont need these anymore
-        settings.on_body = nil
-        settings.on_chunk_complete = nil
-        settings.on_url = nil
-        settings.on_status = nil
-        settings.on_chunk_header = nil
-        settings.on_chunk_complete = nil
-        settings.on_header_field = nil
-        settings.on_header_value = nil
-        settings.on_message_begin = nil
-    }
+    // Decode HTTP until there is nothing more to decode.
+    private func decodeHTTP(ctx: ChannelHandlerContext) throws {
+        // We need to refetch the cumulationBuffer on each loop as it may has changed due re-entrance calls of channelRead(...)
+        while let bufferSlice = self.cumulationBuffer, bufferSlice.readableBytes > 0 {
+            let result = try bufferSlice.withVeryUnsafeBytes { (pointer) -> size_t in
```

Comment:
Why does this use `withVeryUnsafeBytes` and then manipulate the pointer, instead of using `withUnsafeReadableBytes`?

Reply:
because it makes things easier when we calculate the readerIndex based on the baseAddress.